goaci.go

```diff
@@ -47,189 +27,401 @@ func (v *StringVector) Set(str string) error {
+	exec := []string{filepath.Join("/", binary)}
+	exec = append(exec, opts.exec...)
+
+	return &types.App{
+		Exec:  exec,
+		User:  "0",
+		Group: "0",
 	}
-	fn := fi[0].Name()
-	debug("found binary: ", fn)
+}
 
-	// Set up rootfs for ACI layout
-	rfs := filepath.Join(acidir, "rootfs")
-	err = os.MkdirAll(rfs, 0755)
+func getVCSLabel(pathsNames *pathsAndNames) (*types.Label, error) {
+	name, value, err := GetVCSInfo(pathsNames.projectPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get VCS info: %v", err)
+	}
+	acname, err := types.NewACName(name)
 	if err != nil {
-		die(err.Error())
+		return nil, fmt.Errorf("Invalid VCS label: %v", err)
 	}
+	return &types.Label{
+		Name:  *acname,
+		Value: value,
+	}, nil
+}
 
-	// Move the binary into the rootfs
-	ep := filepath.Join(rfs, fn)
-	err = os.Rename(filepath.Join(gobin, fn), ep)
+func prepareManifest(opts *options, pathsNames *pathsAndNames, binary string) (*schema.ImageManifest, error) {
+	name, err := types.NewACName(pathsNames.imageACName)
+	// TODO(jonboulle): could this ever actually happen?
+	if err != nil {
+		return nil, err
+	}
+
+	app := getApp(opts, binary)
+
+	vcsLabel, err := getVCSLabel(pathsNames)
 	if err != nil {
-		die(err.Error())
+		return nil, err
+	}
+	labels := types.Labels{
+		*vcsLabel,
+	}
+
+	manifest := schema.BlankImageManifest()
+	manifest.Name = *name
+	manifest.App = app
+	manifest.Labels = labels
+	return manifest, nil
+}
+
+func copyAssets(opts *options, pathsNames *pathsAndNames) error {
+	placeholderMapping := map[string]string{
+		"<PROJPATH>": pathsNames.projectPath,
+		"<GOPATH>":   pathsNames.goPath,
+	}
+	if err := PrepareAssets(opts.assets, pathsNames.rootFSPath, placeholderMapping); err != nil {
+		return err
 	}
-	debug("moved binary to:", ep)
+	return nil
+}
+
+func moveBinaryToRootFS(pathsNames *pathsAndNames, binary string) error {
+	// Move the binary into the rootfs
+	ep := filepath.Join(pathsNames.rootFSPath, binary)
+	if err := os.Rename(filepath.Join(pathsNames.goBinPath, binary), ep); err != nil {
+		return err
+	}
+	Debug("moved binary to: ", ep)
+	return nil
+}
 
-	exec := []string{filepath.Join("/", fn)}
-	exec = append(exec, execOpts...)
-	// Build the ACI
-	im := schema.ImageManifest{
-		ACKind:    types.ACKind("ImageManifest"),
-		ACVersion: schema.AppContainerVersion,
-		Name:      *name,
-		App: &types.App{
-			Exec:  exec,
-			User:  "0",
-			Group: "0",
-		},
+func writeACI(pathsNames *pathsAndNames, manifest *schema.ImageManifest) error {
+	mode := os.O_CREATE | os.O_WRONLY | os.O_TRUNC
+	of, err := os.OpenFile(pathsNames.imageFileName, mode, 0644)
+	if err != nil {
+		return fmt.Errorf("Error opening output file: %v", err)
 	}
-	debug(im)
+	defer of.Close()
 
 	gw := gzip.NewWriter(of)
+	defer gw.Close()
+
 	tr := tar.NewWriter(gw)
+	defer tr.Close()
+
+	// FIXME: the files in the tar archive are added with the
+	// wrong uid/gid. The uid/gid of the aci builder leaks in the
+	// tar archive. See: #16
```

Comment:
I meant to write the full URL in the comment but github converted my comment in a short link.
``https://github.com/appc/goaci/issues/16``